src/Nethermind/Nethermind.Trie/BatchedTrieVisitor.cs

```diff
@@ -0,0 +1,360 @@
+                trieNode.ResolveNode(_resolver);
+                Interlocked.Increment(ref _activeJobs);
+                trieNode.AcceptResolvedNode(_visitor, _resolver, ctx, recursiveResult);
+                QueueNextNodes(recursiveResult);
+                recursiveResult.Dispose();
+                continue;
+            }
+
+            ValueKeccak keccak = trieNode.Keccak;
+            int partitionIdx = CalculatePartitionIdx(keccak);
+            Interlocked.Increment(ref _activeJobs);
+            Interlocked.Increment(ref _queuedJobs);
+
+            var theStack = _partitions[partitionIdx];
+            lock (theStack)
+            {
+                theStack.Push(new Job(keccak, ctx));
+            }
+        }
+
+        Interlocked.Decrement(ref _activeJobs);
+    }
+
+
+    private void BatchedThread()
+    {
+        ArrayPoolList<(TrieNode, SmallTrieVisitContext)>? currentBatch;
+        ArrayPoolList<(TrieNode, SmallTrieVisitContext)> nextToProcesses = new(_maxBatchSize);
+        ArrayPoolList<int> resolveOrdering = new(_maxBatchSize);
+        while ((currentBatch = GetNextBatch()) != null)
+        {
+            // Storing the idx separately as the ordering is important to reduce memory (approximate dfs ordering)
+            // but the path ordering is important for read amplification
+            resolveOrdering.Clear();
+            for (int i = 0; i < currentBatch.Count; i++)
+            {
+                TrieNode cur = currentBatch[i].Item1;
+
+                cur.ResolveKey(_resolver, false);
+
+                SmallTrieVisitContext ctx = currentBatch[i].Item2;
+
+                if (cur.FullRlp != null) continue;
+                if (cur.Keccak is null) throw new TrieException($"Unable to resolve node without Keccak. ctx: {ctx.Level}, {ctx.ExpectAccounts}, {ctx.IsStorage}, {ctx.BranchChildIndex}");
+
+                resolveOrdering.Add(i);
+            }
+
+            // This innocent looking sort is surprisingly effective when batch size is large enough. The sort itself
+            // take about 0.1% of the time, so not very cpu intensive in this case.
+            resolveOrdering
+                .AsSpan()
+                .Sort((item1, item2) => currentBatch[item1].Item1.Keccak.CompareTo(currentBatch[item2].Item1.Keccak));
+
+            // This loop is about 60 to 70% of the time spent. If you set very high memory budget, this drop to about 50MB.
+            for (int i = 0; i < resolveOrdering.Count; i++)
+            {
+                int idx = resolveOrdering[i];
+
+                (TrieNode nodeToResolve, SmallTrieVisitContext ctx) = currentBatch[idx];
+                try
+                {
+                    Keccak theKeccak = nodeToResolve.Keccak;
+                    nodeToResolve.ResolveNode(_resolver);
+                    nodeToResolve.Keccak = theKeccak; // The resolve may set a key which clear the keccak
```

Comment:
Keccak disappearing could cause issues for logic that saw it was there; can a flag be passed to make sure it doesn't get cleared by this?